### PR TITLE
Use `compileJava` `options.release`

### PR DESCRIPTION
**Pull request by bondolo**
Motivation:
Using `sourceCompatibility` and `targetCompatibility` for specifying
Java version does not always produce correct module metadata because
the setting is quite sensitive to order of operations.
Modifications:
Use `options.release` whenever build JDK is >= Java 9. To support `javac`
compilation using the `--release` option it is necessary to hide direct use
of the `sun.misc.Unsafe` class as this class is not provided in the symbols
used for compiling.
Result:
Correct Java variant versions exported in module files

### CODE REVIEW

File: gradle.properties
Comment (bondolo) [inline gradle.properties:21]: This change is not strictly required.

It disables Gradle from automatically downloading of JDK versions for use for building. When disabled all of the required Java toolchain must be present in the environment, which is what we expect.

**Comment by bondolo** [comment]
I am willing to move the `sun.misc.Unsafe` work out to a separate PR that would be completed first.

File: servicetalk-buffer-api/build.gradle
Comment (idelpivnitskiy) [inline servicetalk-buffer-api/build.gradle:36]: Instead of adding this code snippet in every `build.gradle`, we should rely on our gradle plugin to set it. The plugin is a single source of truth for consistent gradle configuration across all modules.
This can be repeated for submodules that do not apply our plugin.

File: servicetalk-concurrent-jdkflow/build.gradle
Comment (idelpivnitskiy) [inline servicetalk-concurrent-jdkflow/build.gradle:24]: Can we keep `sourceCompatibility` and `targetCompatibility` too? IIRC, some other tools, like IDE, look at them to generate project meta-data correctly.

File: servicetalk-gradle-plugin-internal/src/main/groovy/io/servicetalk/gradle/plugin/internal/ServiceTalkLibraryPlugin.groovy
Comment (idelpivnitskiy) [inline servicetalk-gradle-plugin-internal/src/main/groovy/io/servicetalk/gradle/plugin/internal/ServiceTalkLibraryPlugin.groovy:60]: Same questions here. I'm hesitant to remove these properties. Your change doesn't look conflicting.

**Review by idelpivnitskiy — CHANGES_REQUESTED** [review]

File: servicetalk-buffer-api/build.gradle
Comment (bondolo) [reply servicetalk-buffer-api/build.gradle:36]: This snippet is only in the `build.gradle` of modules which add `testFixtures`. Unless the test fixtures are added then the `compileTestFixturesJava` task is undefined which means this snippet can't be added to the plugin. I was unable, thus far, to find a way to make the behaviour conditional so that happened at the right time and only for the projects with a `compileTestFixturesJava` task.

File: servicetalk-concurrent-jdkflow/build.gradle
Comment (bondolo) [reply servicetalk-concurrent-jdkflow/build.gradle:24]: Sure, as long as it does not cause problems. My reading of https://docs.gradle.org/current/userguide/building_java_projects.html#sec:java_cross_compilation

was that these had been superseded by `options.release` and were no longer required.

Comment (bondolo) [reply servicetalk-concurrent-jdkflow/build.gradle:24]: > Using a toolchain makes it illegal to configure the `sourceCompatibility` or `targetCompatibility` at the java { } extension level.

It appears that these properties are partially obsolete and may eventually be deprecated. I had looked at using `toolchain` rather than `options.release` but it doesn't quite do what we need.

File: servicetalk-buffer-api/build.gradle
Comment (idelpivnitskiy) [reply servicetalk-buffer-api/build.gradle:36]: Gradle allows searching for tasks by their string name: `tasks.findByName("someName")`. It helps to apply logic conditionally. There are a few examples inside our plugin.

File: servicetalk-concurrent-jdkflow/build.gradle
Comment (idelpivnitskiy) [reply servicetalk-concurrent-jdkflow/build.gradle:24]: Until they are officially deprecated, let's keep them. It doesn't hurt to have some redundancy. Those options have 1-1 translation to pom.xml properties and may be useful for IDE.

Comment (bondolo) [reply servicetalk-concurrent-jdkflow/build.gradle:24]: I am fine with keeping them as long as they don't cause problems.

**Comment by bondolo** [comment]
This will have to wait for PR #2303 to merge.

**Comment by bondolo** [comment]
Marked as draft until PR #2303 is merged.

File: gradle.properties
Comment (idelpivnitskiy) [reply gradle.properties:21]: Should we set the same property for `docs` subproject?

**Review by idelpivnitskiy — APPROVED** [review]

File: gradle.properties
Comment (bondolo) [reply gradle.properties:21]: yes, I would argue for setting it in all of our gradle.properties files to prevent downloading and building with random JDK installations.

**Pull request merged**
